File: 6/part1.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

void print_matrix(const std::vector<std::vector<char>>& matrix) {
    for (const auto& row : matrix) {
        for (const auto& cell : row) {
            std::cout << cell;
        }
        std::cout << std::endl;
    }
}

bool move_until_obstacle(std::vector<std::vector<char>>& matrix, std::vector<int>& pos, int direction) {
    try {
        if (direction == 0) { // up
            while (matrix.at(pos[0]).at(pos[1]) != '#') {
                matrix.at(pos[0]).at(pos[1]) = 'X';
                pos[0] -= 1;
                // std::cout << "Guard moving up " << pos[0] << ", " << pos[1] << std::endl;
            }
            pos[0] += 1;
        } else if (direction == 1) { // right
            while (matrix.at(pos[0]).at(pos[1]) != '#') {
                matrix.at(pos[0]).at(pos[1]) = 'X';
                pos[1] += 1;
                // std::cout << "Guard moving right " << pos[0] << ", " << pos[1] << std::endl;
            }
            pos[1] -= 1;
        } else if (direction == 2) { // down
            while (matrix.at(pos[0]).at(pos[1]) != '#') {
                matrix.at(pos[0]).at(pos[1]) = 'X';
                pos[0] += 1;
                // std::cout << "Guard moving down " << pos[0] << ", " << pos[1] << std::endl;
            }
            pos[0] -= 1;
        } else if (direction == 3) { // left
            while (matrix.at(pos[0]).at(pos[1]) != '#') {
                matrix.at(pos[0]).at(pos[1]) = 'X';
                pos[1] -= 1;
                // std::cout << "Guard moving left " << pos[0] << ", " << pos[1] << std::endl;
            }
            pos[1] += 1;
        }
    } catch (const std::out_of_range&) {
        std::cout << "Guard exited the matrix" << std::endl;
        return true;
    }

    // std::cout << "Guard hit an obstacle" << std::endl;
    return false;
}

void func() {
    std::vector<std::vector<char>> matrix;
    std::vector<int> init_pos;
    std::ifstream infile("input.txt");
    std::string line;

    while (std::getline(infile, line)) {
        matrix.push_back(std::vector<char>(line.begin(), line.end()));
        if (init_pos.empty()) {
            for (size_t i = 0; i < line.size(); ++i) {
                if (line[i] == '^') {
                    init_pos = {static_cast<int>(matrix.size()) - 1, static_cast<int>(i)};
                }
            }
        }
    }

    print_matrix(matrix);
    std::cout << "Initial position: " << init_pos[0] << ", " << init_pos[1] << std::endl;

    bool exited = false;
    int direction = 0;

    int counter = 0;
    while (!exited) {
        exited = move_until_obstacle(matrix, init_pos, direction);
        direction = (direction + 1) % 4;
        std::cout << "Guard moved to [" << init_pos[0] << ", " << init_pos[1] << "]" << std::endl;
    }

    // int counter = 0;
    for (const auto& row : matrix) {
        for (const auto& cell : row) {
            if (cell == 'X') {
                counter += 1;
            }
        }
    }

    std::cout << "Guard moved " << counter << " steps" << std::endl;
}

int main() {
    func();
    return 0;
}

File: 6/part1.py
def print_matrix(matrix):
    # return
    for row in matrix:
        print(''.join(row))
    # with open('output.txt', 'w') as f:
        # for row in matrix:
            # f.write(''.join(row) + '\n')

def move_until_obstacle(matrix, pos, direction):
    # print("Guard is in", pos, "moving", direction)
    if direction == 0: # up
        while matrix[pos[0]][pos[1]] != '#':
            matrix[pos[0]][pos[1]] = 'X'
            pos[0] -= 1
            if pos[0] < 0:
                return True
        pos[0] += 1
    elif direction == 1: # right
        while matrix[pos[0]][pos[1]] != '#':
            matrix[pos[0]][pos[1]] = 'X'
            pos[1] += 1
            if pos[1] >= len(matrix[pos[0]]):
                return True
        pos[1] -= 1
    elif direction == 2: # down
        while matrix[pos[0]][pos[1]] != '#':
            matrix[pos[0]][pos[1]] = 'X'
            pos[0] += 1
            if pos[0] >= len(matrix):
                return True
        pos[0] -= 1
    elif direction == 3: # left
        while matrix[pos[0]][pos[1]] != '#':
            matrix[pos[0]][pos[1]] = 'X'
            pos[1] -= 1
            if pos[1] < 0:
                return True
        pos[1] += 1
    
    return False


def func():
    matrix = []
    init_pos = []
    with open('input.txt') as f:
        for line in f:
            matrix.append(list(line.strip()))

            if len(init_pos) == 0:
                for i in range(len(line)):
                    if line[i] == '^':
                        init_pos = [len(matrix)-1, i]
            
    if len(matrix[-1]) == 0:
        matrix.pop()
    print_matrix(matrix)

    exited = False
    direction = 0

    while not exited:
        exited = move_until_obstacle(matrix, init_pos, direction)
        direction = (direction + 1) % 4
        print("Guard moved to", init_pos)

    counter = 0
    for row in matrix:
        for cell in row:
            if cell == 'X':
                counter += 1

    print("Guard moved of", counter, "steps")


if __name__ == '__main__':
    func()